File: SDAC/Esercitazioni/Esercitazione 3/c-source/max_gap.c
#include <assert.h>
#include <stdio.h>
#include <math.h>

#include "max_gap.h"

int maxGap(int* array, int start, int end) {
	// TODO: Da completare
	int* temp = array;
	int medio = (end+start)/2;
	if(end-start == 0) {
		return 0;
	}
	if(end-start == 1) {
		return abs(array[end]-array[start]);
	}
	int a = maxGap(array, start, medio);
	int b;
	if(end == 9) 
		b = maxGap(array, medio, end-1);
	else 
		b = maxGap(array, medio, end);
	if(a>b) return a;
	else return b;
}

File: Strutture Dati/Esercitazioni/Esercitazione 3/java-source/MaxGap.java
public class MaxGap {

    public static int maxGap(int[] array, int start, int end) {
        if(array == null || start < 0 || start > end) {
			throw new RuntimeException("Invalidi array, start o end");
		}
		int size = end - start;

		if(size < 2) {
			return -1;
		}
		if(size == 2) {
			return array[start+1] - array[start];
		}
		int med = start + (size/2);

		int sx = maxGap(array, start, med+1);
		int dx = maxGap(array, med, end);

		return (sx > dx) ? sx : dx;
    }
}
